### PR TITLE
User manager in meta

### PR DESCRIPTION
**Pull request by boshengchen**
Instead of PR #257 .
1. Addressed comments.
2. storage the data using dataman .

### CODE REVIEW

File: src/meta/processors/AuthenticationProcessor.cpp [hunk 1]
```diff
@@ -0,0 +1,310 @@
+/* Copyright (c) 2018 vesoft inc. All rights reserved.
```
Comment (darionyaphet) [inline src/meta/processors/AuthenticationProcessor.cpp]: please update to `2019`

Comment (boshengchen) [reply src/meta/processors/AuthenticationProcessor.cpp]: Sorry, my bad, let me check all new files.

File: src/interface/meta.thrift [hunk 2]
```diff
@@ -77,6 +89,22 @@ struct EdgeItem {
...
 }
 
+
+
+struct UserItem {
+    1: string account,
+    2: string first_name,
+    3: string last_name,
+    4: string email,
+    5: string phone,
```
Comment (darionyaphet) [inline src/interface/meta.thrift]: why supported `phone` ?

File: src/meta/processors/AuthenticationProcessor.cpp (hunk 1 above)
Comment (boshengchen) [reply src/meta/processors/AuthenticationProcessor.cpp]: I have checked the license head, and all the files are 2018 currently. Do we need to be changed them to 2019?

File: src/interface/meta.thrift (hunk 2 above)
Comment (boshengchen) [reply src/interface/meta.thrift]: > why supported `phone` ?

It's just a user property. It is optional. More such options may be added in the future.

File: src/meta/processors/AuthenticationProcessor.cpp (hunk 1 above)
Comment (boshengchen) [reply src/meta/processors/AuthenticationProcessor.cpp]: Changed to 2019 with all new files as below :
src/meta/processors/AuthenticationProcessor.cpp
src/meta/processors/AuthenticationProcessor.h
src/meta/test/AuthProcessorTest.cpp

File: src/interface/meta.thrift (hunk 2 above)
Comment (darionyaphet) [reply src/interface/meta.thrift]: marked as `optional` will be better ?

Comment (boshengchen) [reply src/interface/meta.thrift]: I think It is safe if set the optional to default (i.e. remove the optional keyword). What are the advantages of setting this parameter to `optional`?

Comment (darionyaphet) [reply src/interface/meta.thrift]: I don't think these attributes are necessary.

Comment (dangleptr) [reply src/interface/meta.thrift]: optional and default are not the same. 
default means the field must be written out, but we could read it or not as needs.  If the filed is not set,  we will read a default value for it. 
optional means if the field is not set,  we will read a nullptr.

In this case,  default and optional both are ok.  All of them are backward compatible. 
But for email and phone,  i think we have no needs for them.  But there are some other options we could support.   See https://dev.mysql.com/doc/refman/8.0/en/create-user.html

File: src/interface/meta.thrift [hunk 3]
```diff
@@ -291,6 +319,63 @@ struct HBReq {
     1: common.HostAddr host,
 }
 
+struct CreateUserReq {
+    1: UserItem user,
+    2: string encoded_pwd,
+    3: bool missing_ok,
```
Comment (dangleptr) [inline src/interface/meta.thrift]: What's this?

File: src/meta/MetaServiceUtils.cpp [hunk 4]
```diff
@@ -301,5 +306,129 @@ cpp2::ErrorCode MetaServiceUtils::alterColumnDefs(std::vector<nebula::cpp2::Colu
...
+    std::string key;
+    key.reserve(64);
+    key.append(kUsersTable.data(), kUsersTable.size());
+    key.append(reinterpret_cast<const char*>(&userId), sizeof(userId));
+    return key;
+}
+
+std::string MetaServiceUtils::userVal(const std::string& password,
+                                      const cpp2::UserItem& userItem,
+                                      std::shared_ptr<const SchemaProviderIf> schema) {
```
Comment (dangleptr) [inline src/meta/MetaServiceUtils.cpp]: Don't use dataman in Meta module.  It is too heavy.

File: src/interface/meta.thrift (hunk 3 above)
Comment (boshengchen) [reply src/interface/meta.thrift]: > What's this?

missing_ok？ If the syntax contains `IF NOT EXISTS` ,  missing_ok should be true, else should be false.

**Comment by boshengchen** [comment]
Thanks for your response @darionyaphet @dangleptr , I have two questions to discuss as below :
1, About user options,  In fact, my original design and implementation was based on mysql ，There are options for default roles at that time. It was changed to what it is now after a couple of iterations, Thanks for your suggestions during the design phase. About option `email` and `phone`, The implementation of the syntax has been merged into the mainline，seems that we need to get a confirmed design again . and then re-implement these confirmed design. So,  Please feel free to let me know if have any good ideas . Let's using issue #190 for recording. 
2, About dataman , what are some good ways to avoid using dataman in module meta?  Please refer to #190  for requirements. Thanks.

**Comment by boshengchen** [comment]
Address some comments as below:
1, Changed user data transport protocol from dataman to thrift. 
2, Removed user property email and phone. 
3, Added new user property lock status, this option is referenced from mysql.

File: src/interface/meta.thrift [hunk 5]
```diff
@@ -89,6 +101,21 @@ struct EdgeItem {
     4: common.Schema        schema,
 }
 
+
+
+struct UserItem {
+    1: string account,
+    2: string first_name,
```
Comment (dangleptr) [inline src/interface/meta.thrift]: Do we need first_name, last_name?   
Why not put password into this structure?

File: src/interface/meta.thrift [hunk 6]
```diff
@@ -89,6 +101,21 @@ struct EdgeItem {
...
+    3: string last_name,
+    4: bool   is_lock,
+}
+
+struct RoleItem {
+    1: string account,
```
Comment (dangleptr) [inline src/interface/meta.thrift]: Why not use userID?

File: src/interface/meta.thrift [hunk 7]
```diff
@@ -300,6 +327,72 @@ struct HBReq {
...
 }
 
+struct CreateUserReq {
+    1: UserItem user,
+    2: string encoded_pwd,
+    3: bool missing_ok,
+}
+
+struct DropUserReq {
+    1: string account,
```
Comment (dangleptr) [inline src/interface/meta.thrift]: If you don't use "userId" in all requests, why we need it?

File: src/interface/meta.thrift [hunk 8]
```diff
@@ -89,6 +101,21 @@ struct EdgeItem {
...
+struct UserItem {
+    1: string account,
+    2: string first_name,
+    3: string last_name,
+    4: bool   is_lock,
+}
+
+struct RoleItem {
+    1: string account,
+    2: string space,
```
Comment (dangleptr) [inline src/interface/meta.thrift]: why not spaceId?

File: src/interface/meta.thrift [hunk 9]
```diff
@@ -43,10 +46,19 @@ enum AlterSchemaOp {
 } (cpp.enum_strict)
 
 
+enum RoleType {
+    GOD    = 0x01,
```
Comment (dangleptr) [inline src/interface/meta.thrift]: Add some comments, what's the meaning for different roles?

File: src/interface/meta.thrift [hunk 10]
```diff
@@ -300,6 +327,72 @@ struct HBReq {
...
+
+struct ListUsersResp {
+    1: ErrorCode code,
+    // Valid if ret equals E_LEADER_CHANGED.
+    2: common.HostAddr  leader,    
+    3: list<UserItem> users,
+}
+
+struct ListRolesReq {
+    1: string space,
```
Comment (dangleptr) [inline src/interface/meta.thrift]: ditto

File: src/interface/meta.thrift (hunk 5 above)
Comment (boshengchen) [reply src/interface/meta.thrift]: Thanks for your review @dangleptr , here are the answers to each question:
1: About first_name and last_name, Discussed with  @sherman-the-tank in issue #190 , Need to add more properties to a user, such as email address, first name, last name, etc.
2:About password is not contained in the structure UserItem，the reason is functional difference is between password and other user attributes. password is only used when connecting or changing password, so password is not normal user attributes. Also, fast read and write do not require serialization or deserialization of structures.

File: src/interface/meta.thrift (hunk 9 above)
Comment (boshengchen) [reply src/interface/meta.thrift]: Good suggest!

**Comment by boshengchen** [comment]
About SpaceID and UserID exists multiple places, I've actually thought about it，Because these two structures are used for user management often. So  can use these names directly . I think convert
 from name to id is superfluous in here. What do you think?
UserID and SpaceID are actually useful, they will be used for privilege management. I will use UserID and SpaceID to implement some caches in meta server or meta client.

File: src/interface/meta.thrift (hunk 5 above)
Comment (dangleptr) [reply src/interface/meta.thrift]: > 1: About first_name and last_name, Discussed with @sherman-the-tank in issue #190 , Need to add more properties to a user, such as email address, first name, last name, etc.

We'd better just saved the information our system used.  Just supposed that, when you create a account,  the system ask you to fill your personal information,  is it reasonable?

> 2:About password is not contained in the structure UserItem，the reason is functional difference is between password and other user attributes. password is only used when connecting or changing password, so password is not normal user attributes. Also, fast read and write do not require serialization or deserialization of structures.

Make sense.

**Comment by dangleptr** [comment]
> About SpaceID and UserID exists multiple places, I've actually thought about it，Because these two structures are used for user management often. So can use these names directly . I think convert
> from name to id is superfluous in here. What do you think?
> UserID and SpaceID are actually useful, they will be used for privilege management. I will use UserID and SpaceID to implement some caches in meta server or meta client.

It make sense for userId,  but not for spaceID.  We'd better use spaceId to join different entries instead of spaceName.

**Comment by dangleptr** [comment]
By the way,  please refer #483 .  Put the user related processors under a separate dir.

**Comment by boshengchen** [comment]
> It make sense for userId, but not for spaceID. We'd better use spaceId to join different entries instead of spaceName.
That makes sense, let me improve the space name and RoleItem related.

**Comment by boshengchen** [comment]
1,Refactor user processors code structure. 
2,Remove user properties first name and last name. 
3, Add new user properties of resource limit. 
4, Changed related thrift structure from `'name'` to `'id'`, 
5,Add some comments for role.

**Comment by boshengchen** [comment]
Ready to review. Please review again. Thanks.

File: src/graph/PermissionManager.h [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+
+#ifndef GRAPH_PERMISSIONMANAGER_H
+#define GRAPH_PERMISSIONMANAGER_H
+
+// Operation and permission define:
```
Comment (dangleptr) [inline src/graph/PermissionManager.h]: Well done

File: src/meta/processors/BaseProcessor.inl [hunk 12]
```diff
@@ -162,6 +162,18 @@ Status BaseProcessor<RESP>::spaceExist(GraphSpaceID spaceId) {
     return Status::SpaceNotFound();
 }
 
+template<typename RESP>
+Status BaseProcessor<RESP>::userExist(UserID spaceId) {
```
Comment (dangleptr) [inline src/meta/processors/BaseProcessor.inl]: Could we unify the userExist with getUserId ?

File: src/meta/processors/partsMan/DropSpaceProcessor.cpp [hunk 13]
```diff
@@ -41,6 +41,16 @@ void DropSpaceProcessor::process(const cpp2::DropSpaceReq& req) {
     deleteKeys.emplace_back(MetaServiceUtils::indexSpaceKey(req.get_space_name()));
     deleteKeys.emplace_back(MetaServiceUtils::spaceKey(spaceId));
 
+    // delete related role data.
```
Comment (dangleptr) [inline src/meta/processors/partsMan/DropSpaceProcessor.cpp]: Now we just delete the parts information when deleting space.  We could delete the useless user and schema when compaction.

File: src/meta/processors/usersMan/AuthenticationProcessor.cpp [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+    LOG(INFO) << "Create User " << user.get_account() << ", userId " << userId;
+    data.emplace_back(MetaServiceUtils::userKey(userId),
+                      MetaServiceUtils::userVal(req.get_encoded_pwd(), user));
+    resp_.set_code(cpp2::ErrorCode::SUCCEEDED);
+    resp_.set_id(to(userId, EntryType::USER));
+    doPut(std::move(data));
+}
+
+
+void AlterUserProcessor::process(const cpp2::AlterUserReq& req) {
```
Comment (dangleptr) [inline src/meta/processors/usersMan/AuthenticationProcessor.cpp]: It seems the alter is to change the whole value. 
I want to know what's the language for alter user?

File: src/meta/processors/usersMan/AuthenticationProcessor.cpp [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+    folly::SharedMutex::WriteHolder wHolder(LockUtils::userLock());
+    auto roleKey = MetaServiceUtils::roleKey(roleItem.get_space_id(), roleItem.get_user_id());
+    resp_.set_id(to(roleItem.get_user_id(), EntryType::USER));
+    resp_.set_code(cpp2::ErrorCode::SUCCEEDED);
+    doRemove(std::move(roleKey));
+}
+
+
+void ChangePasswordProcessor::process(const cpp2::ChangePasswordReq& req) {
+    folly::SharedMutex::WriteHolder wHolder(LockUtils::userLock());
```
Comment (dangleptr) [inline src/meta/processors/usersMan/AuthenticationProcessor.cpp]: Change pwd is a special sentence?

File: src/meta/processors/usersMan/AuthenticationProcessor.cpp [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+void CheckPasswordProcessor::process(const cpp2::CheckPasswordReq& req) {
+    folly::SharedMutex::ReadHolder rHolder(LockUtils::userLock());
+    auto userRet = getUserId(req.get_account());
+    if (!userRet.ok()) {
+        resp_.set_code(to(userRet.status()));
+        onFinished();
+        return;
+    }
+
+    if (!checkPassword(userRet.value(), req.get_encoded_pwd())) {
```
Comment (dangleptr) [inline src/meta/processors/usersMan/AuthenticationProcessor.cpp]: I want to know when to check the password?

File: src/meta/processors/BaseProcessor.inl (hunk 12 above)
Comment (boshengchen) [reply src/meta/processors/BaseProcessor.inl]: `userExist` is judged by UserID , But  `getUserId` is  judged by userName. And they query for different tables userIndex and userData. They are same with spaceExist and getSpaceId. I think it's best not to unify them. WDYT?

File: src/meta/processors/partsMan/DropSpaceProcessor.cpp (hunk 13 above)
Comment (boshengchen) [reply src/meta/processors/partsMan/DropSpaceProcessor.cpp]: Yeah, that makes sense.  Let me modify it later.

File: src/meta/processors/usersMan/AuthenticationProcessor.cpp (hunk 14 above)
Comment (boshengchen) [reply src/meta/processors/usersMan/AuthenticationProcessor.cpp]: No, Only change properties that require alter.

**Alter user sentences:**
```
ALTER USER account WITH with_user_opt_list

with_user_opt_item
    : KW_ACCOUNT KW_LOCK
    | KW_ACCOUNT KW_UNLOCK
    | KW_MAX_QUERIES_PER_HOUR INTEGER
    | KW_MAX_UPDATES_PER_HOUR INTEGER
    | KW_MAX_CONNECTIONS_PER_HOUR INTEGER
    | KW_MAX_USER_CONNECTIONS INTEGER
    ;

with_user_opt_list
    : with_user_opt_item
    | with_user_opt_list COMMA with_user_opt_item
    ;
```

File: src/meta/processors/usersMan/AuthenticationProcessor.cpp (hunk 15 above)
Comment (boshengchen) [reply src/meta/processors/usersMan/AuthenticationProcessor.cpp]: Yes, Change password is a separate sentence. No need to verify the password if it is a `GOD` role. else the password must be validated. And `GOD` can change anyone's password. Other roles can only change their own password.

```
change_password_sentence
    : CHANGE PASSWORD account TO new_password
    | CHANGE PASSWORD account FROM old_password TO new_password
    ;
```

File: src/meta/processors/usersMan/AuthenticationProcessor.cpp (hunk 16 above)
Comment (boshengchen) [reply src/meta/processors/usersMan/AuthenticationProcessor.cpp]: This is a reserved interface，Used for client or terminal connections.

**Comment by boshengchen** [comment]
To improve the drop space , don't delete useless role data.

File: src/meta/processors/BaseProcessor.inl (hunk 12 above)
Comment (dangleptr) [reply src/meta/processors/BaseProcessor.inl]: Fair enough.

**Comment by dangleptr** [comment]
Jenkins go

**Comment by boshengchen** [comment]
1, Rebased 
2, Resolved compilation errors with depends on  `raftex_obj raftex_thrift_obj wal_obj time_obj`.
Please review again . Thanks.

**Comment by boshengchen** [comment]
Rebased and resolved some conflicts.

File: src/meta/processors/usersMan/AuthenticationProcessor.cpp [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+        onFinished();
+        return;
+    }
+    std::vector<std::string> keys;
+    keys.emplace_back(MetaServiceUtils::indexUserKey(req.get_account()));
+    keys.emplace_back(MetaServiceUtils::userKey(ret.value()));
+
+    // Collect related roles by user.
+    std::unique_ptr<kvstore::KVIterator> iter;
+    std::string prefix = "__roles__";
```
Comment (laura-ding) [inline src/meta/processors/usersMan/AuthenticationProcessor.cpp]: I think the prefix should be used kRolesTable although it define in cpp file

File: src/meta/test/CMakeLists.txt [hunk 18]
```diff
@@ -171,3 +171,33 @@ nebula_link_libraries(
...
+        AuthProcessorTest.cpp
+        $<TARGET_OBJECTS:meta_service_handler>
+        $<TARGET_OBJECTS:kvstore_obj>
+        $<TARGET_OBJECTS:meta_client>
+        $<TARGET_OBJECTS:meta_thrift_obj>
+        $<TARGET_OBJECTS:common_thrift_obj>
+        $<TARGET_OBJECTS:raftex_obj>
+        $<TARGET_OBJECTS:raftex_thrift_obj>
+        $<TARGET_OBJECTS:wal_obj>
+	$<TARGET_OBJECTS:thrift_obj>
```
Comment (laura-ding) [inline src/meta/test/CMakeLists.txt]: Alignment ：185L and 189L

File: src/meta/test/TestUtils.h [hunk 19]
```diff
@@ -184,6 +185,37 @@ class TestUtils {
 
         return sc;
     }
+
+    static StatusOr<UserID> createUser(kvstore::KVStore* kv,
```
Comment (laura-ding) [inline src/meta/test/TestUtils.h]: Strange alignment

File: src/meta/processors/usersMan/AuthenticationProcessor.cpp (hunk 17 above)
Comment (boshengchen) [reply src/meta/processors/usersMan/AuthenticationProcessor.cpp]: Good ! We are think alike, I don't feel comfortable using it here either. Let me try.

File: src/meta/test/CMakeLists.txt (hunk 18 above)
Comment (boshengchen) [reply src/meta/test/CMakeLists.txt]: Sharp and bright eyes

**Comment by boshengchen** [comment]
Address laura-ding's comment

File: src/meta/test/CMakeLists.txt [hunk 20]
```diff
@@ -182,11 +182,11 @@ add_executable(
         $<TARGET_OBJECTS:raftex_obj>
         $<TARGET_OBJECTS:raftex_thrift_obj>
         $<TARGET_OBJECTS:wal_obj>
-	$<TARGET_OBJECTS:thrift_obj>
+	    $<TARGET_OBJECTS:thrift_obj>
```
Comment (laura-ding) [inline src/meta/test/CMakeLists.txt]: I see it's still not aligned

Comment (boshengchen) [reply src/meta/test/CMakeLists.txt]: My bad, resolved.  Thanks for your careful review @laura-ding

**Comment by dangleptr** [comment]
Jenkins go

File: src/meta/test/CMakeLists.txt [hunk 21]
```diff
@@ -171,3 +171,33 @@ nebula_link_libraries(
     gtest
 )
 nebula_add_test(meta_http_test)
+add_executable(
```
Comment (laura-ding) [inline src/meta/test/CMakeLists.txt]: image
There are no aligned with 4 Spaces

Comment (boshengchen) [reply src/meta/test/CMakeLists.txt]: A stupid mistake，corrected.

**Comment by boshengchen** [comment]
Jenkins go

**Comment by dangleptr** [comment]
Jenkins go

**Pull request merged**